File: apps/server/agentcore/api/routes/conversations/browser_sessions.py
"""Browser session list/create/close/navigate — multi ``session_id`` (M0 / M2).

Owner-only. Conversation-level thin wrap: the registry is the single in-memory map
(``session_id`` primary key); these endpoints expose list / create / close / navigate
without a second Registry.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from agentcore.api.dependencies import AuthUser, get_db
from agentcore.api.schemas import (
    BrowserSessionCreateRequest,
    BrowserSessionListResponse,
    BrowserSessionNavigateRequest,
    BrowserSessionNavPatch,
    BrowserSessionView,
    StatusResponse,
)
from agentcore.config import settings
from agentcore.conversation.common import resolve_turn_file_workspace
from agentcore.core.errors import NotFoundError, ValidationError
from agentcore.db.models import Conversation
from agentcore.db.repositories import ConversationRepository
from agentcore.folders.placement import resolve_folder_placement
from agentcore.runtime.browser import default_browser_session_registry
from agentcore.tools.sandbox.browser.protocol import BrowserCommand, BrowserSessionRequest
from agentcore.workspace.locate import workspace_root_path

from ._helpers import _get_owned_conversation, _require_owned_conversation

router = APIRouter(prefix="/conversations", tags=["conversations"])


async def _conversation_workspace_root_str(
    conv: Conversation, user_id: str, session: AsyncSession
) -> str | None:
    """Same physical root as file tools. Do not mkdir for opening a browser."""
    ws_folder_id, _ = resolve_turn_file_workspace(
        birth_folder_id=conv.folder_id,
        auto_desk_folder_id=getattr(conv, "auto_desk_folder_id", None),
    )
    placement = await resolve_folder_placement(ws_folder_id, session=session)
    try:
        root = workspace_root_path(
            user_id=user_id,
            folder_rel_path=placement.rel_path,
            conversation_id=conv.id,
        )
    except ValueError:
        return None
    if not root.is_dir():
        return None
    return str(root)


def _view_from_info(info) -> BrowserSessionView:
    return BrowserSessionView(
        session_id=info.session_id,
        conversation_id=info.conversation_id,
        host_kind=info.host_kind,
        run_id=info.run_id,
        control=info.control,
        created_at=info.created_at,
        last_used=info.last_used,
        url=getattr(info, "url", None),
        title=getattr(info, "title", None),
    )


@router.get("/{conversation_id}/browser/sessions", response_model=BrowserSessionListResponse)
async def list_browser_sessions(
    conversation_id: str,
    user: AuthUser,
    session: AsyncSession = Depends(get_db),
) -> BrowserSessionListResponse:
    """List live browser sessions for the conversation (owner-only)."""
    await _require_owned_conversation(
        conversation_id, user.user_id, ConversationRepository(session)
    )
    reg = default_browser_session_registry()
    infos = reg.list_by_conversation(conversation_id)
    active = reg.resolve_session_id(conversation_id)
    return BrowserSessionListResponse(
        data=[_view_from_info(i) for i in infos],
        active_session_id=active,
    )


@router.post(
    "/{conversation_id}/browser/sessions",
    response_model=BrowserSessionView,
    status_code=201,
)
async def create_browser_session(
    conversation_id: str,
    body: BrowserSessionCreateRequest,
    user: AuthUser,
    session: AsyncSession = Depends(get_db),
) -> BrowserSessionView:
    """Create a new browser session tab (owner-only)."""
    conv = await _get_owned_conversation(
        conversation_id, user.user_id, ConversationRepository(session)
    )
    workspace_root = await _conversation_workspace_root_str(conv, user.user_id, session)
    if body.host_kind == "sandbox" and not workspace_root:
        raise ValidationError("云端浏览器需要已挂载的工作区盘（禁止无盘 jail）。")
    reg = default_browser_session_registry()
    request = BrowserSessionRequest(
        conversation_id=conversation_id,
        workspace_root=workspace_root,
        viewport_width=int(settings.browser_keyframe_width),
        jpeg_quality=int(settings.browser_keyframe_jpeg_quality),
        host_kind=body.host_kind,
    )
    _browser, _kf, session_id = await reg.create(
        request,
        host_kind=body.host_kind,
        activate=body.activate,
    )
    infos = [i for i in reg.list_by_conversation(conversation_id) if i.session_id == session_id]
    return _view_from_info(infos[0])


@router.delete(
    "/{conversation_id}/browser/sessions/{session_id}",
    response_model=StatusResponse,
)
async def close_browser_session(
    conversation_id: str,
    session_id: str,
    user: AuthUser,
    session: AsyncSession = Depends(get_db),
) -> StatusResponse:
    """Close one browser session by id (owner-only)."""
    await _require_owned_conversation(
        conversation_id, user.user_id, ConversationRepository(session)
    )
    reg = default_browser_session_registry()
    if reg.conversation_of(session_id) != conversation_id:
        raise NotFoundError("浏览器会话不存在")
    await reg.close_session(session_id)
    return StatusResponse()


@router.patch(
    "/{conversation_id}/browser/sessions/{session_id}",
    response_model=BrowserSessionView,
)
async def patch_browser_session_nav(
    conversation_id: str,
    session_id: str,
    body: BrowserSessionNavPatch,
    user: AuthUser,
    session: AsyncSession = Depends(get_db),
) -> BrowserSessionView:
    """L7 最小：回写 url/title（用户地址栏 / Bridge 导航后）。"""
    await _require_owned_conversation(
        conversation_id, user.user_id, ConversationRepository(session)
    )
    reg = default_browser_session_registry()
    if reg.conversation_of(session_id) != conversation_id:
        raise NotFoundError("浏览器会话不存在")
    if not reg.update_nav(session_id, url=body.url, title=body.title):
        raise NotFoundError("浏览器会话不存在")
    infos = [i for i in reg.list_by_conversation(conversation_id) if i.session_id == session_id]
    return _view_from_info(infos[0])


@router.post(
    "/{conversation_id}/browser/sessions/{session_id}/navigate",
    response_model=BrowserSessionView,
)
async def navigate_browser_session(
    conversation_id: str,
    session_id: str,
    body: BrowserSessionNavigateRequest,
    user: AuthUser,
    session: AsyncSession = Depends(get_db),
) -> BrowserSessionView:
    """Owner address-bar navigate: registry ``send(BrowserCommand navigate)``.

    Works for sandbox (Web) and local host_kind sessions alike.
    """
    await _require_owned_conversation(
        conversation_id, user.user_id, ConversationRepository(session)
    )
    reg = default_browser_session_registry()
    if reg.conversation_of(session_id) != conversation_id:
        raise NotFoundError("浏览器会话不存在")
    browser = reg.get(session_id)
    if browser is None:
        raise NotFoundError("浏览器会话不存在")
    result = await browser.send(BrowserCommand(action="navigate", args={"url": body.url}))
    if not result.ok:
        raise ValidationError(result.error or "浏览器导航失败")
    final_url = (result.data or {}).get("final_url")
    title = (result.data or {}).get("title")
    reg.update_nav(
        session_id,
        url=str(final_url) if final_url is not None else body.url,
        title=str(title) if title is not None else None,
    )
    infos = [i for i in reg.list_by_conversation(conversation_id) if i.session_id == session_id]
    if not infos:
        raise NotFoundError("浏览器会话不存在")
    return _view_from_info(infos[0])

File: apps/desktop/src/renderer/services/browserSessions.ts
import { api } from "@/services/api";
import {
  getActiveSidecarTarget,
  resolveConversationLocalTarget,
  resolveSidecarRoot,
} from "@/services/sidecarRouting";

/**
 * 云端浏览器会话 list / create / navigate / close（对齐 `…/browser/sessions`）。
 *
 * Local hydrate：能解析 sidecar target 时走 ``sidecarApi.listBrowserSessions``
 *（同进程 Registry）；本机绑定会话但走不了 sidecar → 空 list（**禁止**云 GET，
 * 云 Registry 对 Local 恒空=假清空）；仅非 Local 绑定才云 GET。
 *
 * 空白页签仍是本地壳态、不自动 POST；Web 地址栏回车才显式
 * {@link createBrowserSession}（sandbox）+ {@link navigateBrowserSession}。
 * 桌面有 `browserApi` 时走 Local WebContents，不经本 create。
 * 走 `services/api` 以复用 401 刷新 / CSRF（与 browserInput 同路）。
 */

export type BrowserHostKind = "sandbox" | "local";
export type BrowserControl = "agent" | "user";

/** 客户端投影（camelCase）。 */
export interface BrowserSessionInfo {
  sessionId: string;
  conversationId: string;
  hostKind: BrowserHostKind;
  control: BrowserControl;
  runId: string | null;
  createdAt: number;
  lastUsed: number;
  url?: string | null;
  title?: string | null;
}

export interface BrowserSessionList {
  sessions: BrowserSessionInfo[];
  activeSessionId: string | null;
}

/** 服务端 wire（snake_case）。 */
interface BrowserSessionWire {
  session_id: string;
  conversation_id: string;
  host_kind: BrowserHostKind;
  control: BrowserControl;
  run_id?: string | null;
  created_at: number;
  last_used: number;
  url?: string | null;
  title?: string | null;
}

interface BrowserSessionListWire {
  data: BrowserSessionWire[];
  active_session_id?: string | null;
}

function sessionsPath(conversationId: string): string {
  return `/v1/conversations/${encodeURIComponent(conversationId)}/browser/sessions`;
}

function sessionPath(conversationId: string, sessionId: string): string {
  return `${sessionsPath(conversationId)}/${encodeURIComponent(sessionId)}`;
}

function fromWire(w: BrowserSessionWire): BrowserSessionInfo {
  return {
    sessionId: w.session_id,
    conversationId: w.conversation_id,
    hostKind: w.host_kind,
    control: w.control,
    runId: w.run_id ?? null,
    createdAt: w.created_at,
    lastUsed: w.last_used,
    url: w.url ?? null,
    title: w.title ?? null,
  };
}

function fromListWire(r: BrowserSessionListWire): BrowserSessionList {
  return {
    sessions: (r.data ?? []).map(fromWire),
    activeSessionId: r.active_session_id ?? null,
  };
}

/** GET …/browser/sessions —— Local→sidecar Registry；本机绑定禁云；否则云端 list。 */
export async function listBrowserSessions(
  conversationId: string,
): Promise<BrowserSessionList> {
  const active = getActiveSidecarTarget(conversationId);
  const target = active ?? (await resolveSidecarRoot(conversationId));
  if (
    target &&
    typeof window !== "undefined" &&
    window.sidecarApi?.listBrowserSessions
  ) {
    const raw = await window.sidecarApi.listBrowserSessions({
      rootId: target.rootId,
      subpath: target.subpath,
      conversationId,
    });
    return fromListWire(raw);
  }
  // 本机绑定会话：云 Registry 无 Local session，GET 恒空会假清空右坞。
  if ((await resolveConversationLocalTarget(conversationId)) != null) {
    return { sessions: [], activeSessionId: null };
  }
  const r = await api.get<BrowserSessionListWire>(sessionsPath(conversationId));
  return fromListWire(r);
}

/** POST …/browser/sessions —— 显式开一页（Web 地址栏 / 壳侧；默认 sandbox）。 */
export function createBrowserSession(
  conversationId: string,
  opts?: { hostKind?: BrowserHostKind; activate?: boolean },
): Promise<BrowserSessionInfo> {
  return api
    .post<BrowserSessionWire>(sessionsPath(conversationId), {
      host_kind: opts?.hostKind ?? "sandbox",
      activate: opts?.activate ?? true,
    })
    .then(fromWire);
}

/** POST …/browser/sessions/{id}/navigate —— 对该 session 发 BrowserCommand navigate。 */
export function navigateBrowserSession(
  conversationId: string,
  sessionId: string,
  url: string,
): Promise<BrowserSessionInfo> {
  return api
    .post<BrowserSessionWire>(
      `${sessionPath(conversationId, sessionId)}/navigate`,
      {
        url,
      },
    )
    .then(fromWire);
}

/** DELETE …/browser/sessions/{session_id} —— 关掉一页服务端会话（会拆 gVisor）。 */
export async function closeBrowserSession(
  conversationId: string,
  sessionId: string,
): Promise<void> {
  await api.delete(sessionPath(conversationId, sessionId));
}

/** PATCH …/browser/sessions/{session_id} —— L7 回写 url/title（地址栏 / 本机导航后）。 */
export async function patchBrowserSessionNav(
  conversationId: string,
  sessionId: string,
  nav: { url?: string | null; title?: string | null },
): Promise<void> {
  await api.patch(sessionPath(conversationId, sessionId), nav);
}
